### DISCUSSION TITLE
RSGI response_file Range header support

### DISCUSSION THREAD

**Discussion by myers** [emmett-framework/granian#157] [Ideas]
Would there be a desire for [http Range header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) support for the `response_file` RSGI response?  

I would like to use this when serving large video files, to allow seeking within the file.

**Comment by gi0baro** [comment]
This might be hard to implement, at the moment I'm not sure we have support for this from the inner libraries used (tokio mainly). But I'm definitely up for PRs investigating this chance.

An alternative approach might be using the [response_stream](https://github.com/emmett-framework/granian/blob/master/docs/spec/RSGI.md#http-protocol-interface) protocol feature and handle the read from Python itself.

**Comment by myers** [comment]
I made an attempt to implement this.  My main problem is that https://github.com/emmett-framework/granian/blob/master/src/rsgi/types.rs#L221 has no access to the `RSGIScope` to be able to read request headers to find the range header.  Just brainstorming maybe `response_file` could change to have another argument that gives the byte range to serve.

**Comment by myers** [comment]
Draft PR: https://github.com/emmett-framework/granian/pull/159

**Comment by matthiask** [comment]
I think Range header support may be interesting to have for ASGI as well. I have been looking into using Starlette's `FileResponse` (https://www.starlette.io/responses/#fileresponse) for static file serving, but it doesn't support the range header yet. I'm wondering if range support could be added to the ASGI pathsend extension, and if maybe the partial file serving could be left to the protocol server. I'm not sure if something like this might also help the RSGI side of things though.

**Reply by gi0baro** [reply]
@matthiask I think ASGI `pathsend` might be the right extension to add this feature; you can probably open up an issue/PR to asgiref and propose the changes.
I'm not sure about the implementation though, as per RSGI it is unclear to me which component should handle the headers: it feels strange to me to have the server modifying the application headers in any circumstance; so I'd probably prefer to have the involved framework/app to manage those.

**Reply by matthiask** [reply]
>  I'm not sure about the implementation though, as per RSGI it is unclear to me which component should handle the headers: it feels strange to me to have the server modifying the application headers in any circumstance; so I'd probably prefer to have the involved framework/app to manage those.

Yes, I agree. I think the framework/app should send back the byte ranges to serve somehow, and the protocol server should handle only the actual loading and sending of the byte ranges. In ASGI terms this could mean an additional key in the `http.response.pathsend` message, in RSGI terms maybe a fourth, optional argument to `response_file`. Not sure. 

> I think ASGI pathsend might be the right extension to add this feature; you can probably open up an issue/PR to asgiref and propose the changes.

Thanks, I'll probably do that in the next (few) weeks!

**Comment by matthewgrossman** [comment]
Trying to follow across a few issues, but it seems that we never exactly resolved this (in either ASGI/RSGI), right?

This seems to be the latest w.r.t ASGI: https://github.com/django/asgiref/issues/469. And latest w.r.t RSGI: https://github.com/emmett-framework/granian/pull/159

Is that right, or has some PR merged somewhere that enables Range header for file GETs in granian?

**Reply by gi0baro** [reply]
#159 will be merged and released in 2.6.

ASGI changes require:
- the implementation
- the protocol documentation update
- pinging any fw out there using `pathsend` to actually use the ranges interface

So, far from "resolved" probably.
btw, you don't need any specific support for "ranges" in Granian with an ASGI application. You simply can't use pathsend for that atm.

**Reply by matthewgrossman** [reply]
Great, thanks for the update!

Agreed that `pathsend` is not necessary for ranged GETs, but for performance gains (e.g having rust handle the file reading + networking) `pathsend` would need to be used, right?

Maybe my mental model is wrong, but here's how I understand it: The improved ASGI pathsend discussed in https://github.com/django/asgiref/issues/469 would allow my ASGI app to return to granian "hey return this file with bytes [X, Y]", and then granian rust-side would handle actual streaming those bytes back to the client. 

This would mean the data wouldn't have to actually flow through the python process at all, which I imagine would result in a large performance improvement?

**Reply by gi0baro** [reply]
Correct.

But, unless you're writing your app in _plain ASGI_, you would need the framework in use to adopt ASGI pathsend to get the performance benefit. AFAIK, the only ASGI framework with that support is Starlette (and thus FastAPI).
All of this to say, at least from my side, that ASGI pathsend extension is not _at the top of my priority list_ :)

**Reply by matthewgrossman** [reply]
Got it, makes sense. Yep that feels pretty far away 😅 Given that we have to first get the protocol/documentation agreed upon, before any ASGI servers (granian) or frameworks (starlette) even bother working on it.

w.r.t implementation: Doesn't IMO look like the starlette changes would be too bad, seems nicely scoped to [here](https://github.com/Kludex/starlette/blob/49d4de92867cb38a781069701ad57cecab4a1a36/starlette/responses.py#L384-L414). And based on the work in https://github.com/emmett-framework/granian/pull/159 I imagine it wouldn't be too bad to just wire the new `pathsend+range` ASGI dict to that same impl (but of course I'd defer to you on that).

I guess next step is to bump that thread in django/asgiref and try to get a proposal through? I see you have experience getting things into ASGI based on this comment https://github.com/django/asgiref/issues/423#issuecomment-2373537065. I'm happy to (attempt to) move things along

**Reply by gi0baro** [reply]
> I guess next step is to bump that thread in django/asgiref and try to get a proposal through? I see you have experience getting things into ASGI based on this comment https://github.com/django/asgiref/issues/423#issuecomment-2373537065. I'm happy to (attempt to) move things along

I'd say anybody willing to add to the spec what I described in https://github.com/django/asgiref/issues/469#issuecomment-2373742600 is completely entitled to do so by opening a PR over there. Eg: this https://github.com/django/asgiref/pull/407 is how `pathsend` was added to the ASGI spec.
